### ISSUE TITLE
X509AttributeCertificateHolder encode and decode

### ISSUE THREAD

**Issue by emreg89** [bcgit/bc-java#196]
Hello,

I tried the X509AttributeCertificateHolder especially to send privileges over the network. A problem appears with the attributes, when the holder is serialized. The origin attribute values are replaced by something else.

I enhanced the existing AttrCertExample with the following two lines of code:

        byte[] arr = att.getEncoded();
        X509AttributeCertificateHolder holder = new X509AttributeCertificateHolder(arr);

The result was, that the attributes changed. I don't know if the failure comes up while encoding or creating a new holder from the encoded byte array.

A simple test is:

        for (int i = 0; i < holder.getAttributes().length; i++) {
            Attribute attr1 = holder.getAttributes()[i];
            Attribute attr2 = att.getAttributes()[i];

            System.out.println("Holder value after decode: " + attr1.getAttrValues());
            System.out.println("Holder value before encode: " + attr2.getAttrValues());
        }

And the result of the test is:

> Holder value after decode: [[[1][6]#69643a2f2f444155313233343536373839]]
> Holder value before encode: [Name: id://DAU123456789 - Auth: N/A]

Normally I would expect that, the holder value before encode and the holder value after decode should be the same.

**Comment by bcgit** [comment]
Okay, I think what is happening here is that the type information is being lost so after decode you are getting back the raw ASN.1 object not the attribute that you originally passed. The objects are still the "same" and I think you will find .equals() works - it's just the toString() method is different. I'll see if I can improve it.

**Comment by bcgit** [comment]
I've had a bit of a play with this, I don't think there's anything we can do sensibly, although that is not to say there is  not anything you can do. The issue is the primitive encoding is the only thing to go on when the certificate is reconstructed, the problem with introducing higher level attribute types is that we can't really make the call as to what things mean, except in a few cases. I'd suggest the best approach is to have a wrapper class which returns the high level attributes you want to see, that way you will be sure that things will always work correctly for what you want to see.

**Comment by Blackjack92** [comment]
This solution is not very satisfaying because when the attribute certificate is send over the network overhead information is needed and the certificate cannot be easily stored. 
Why does it work, when a X509Certificate is transmitted? 
Why is the X509AttributeCertificate deprecated, this would be the perfect approach for sending the attribute certificate?

**Comment by bcgit** [comment]
I'm not sure I understand this comment, it has zero bearing on the issue described above. A lot of attributes are proprietary and it is trivial to store an attribute certificate via getEncoded().

**Comment by Blackjack92** [comment]
I am sorry for my harsh words :) I meant why it is necessary to use a holder - and the X509AttributeCertificate is deprecated?
Because for a Public Key Certificate a X509Certificate object does exist. The same would be good for Attribute Certificates because I don't want to transmit a holder object if it is possible to transmit a Attribute Certificate object.

**Comment by bcgit** [comment]
Okay, this I can answer. The first thing was we needed to get as much of the certificate related stuff out of the provider jar as possible. The issue there was the provider needs to be signed by a specific signature, the other APIs do not. In the case of FIPS this issue become even worse as it would have meant the attribute certificate support was within the boundary of the FIPS Java module. If you don't know what it takes to make a change in a context like that, you really don't want to. After, well, we had X509CertificateHolder so we just followed suite. At this point, the holder class is an attribute certificate.
